feat: Add loongarch64 nightly and release builds

yzewei opened the pull request:
## Summary
<!-- Mandatory
Explain here the why, the rationale and motivation, for the changes.
-->
Provide loongarch support
The LoongArch architecture (LoongArch) is an Instruction Set Architecture (ISA) that has a RISC style.
## Checklist
<!-- Mandatory
Please confirm the following by replacing the space with an "x" between the []:
-->
- [x] Wrote appropriate unit tests.
- [x] No AI generated code was used in this PR

## Related issues
<!-- Mandatory
All PRs should resolve an issue, if one does not exist, please open one.
-->
resolves　#13798

yzewei commented:
!signed-cla

yzewei commented:
!signed-cla

powersj reviewed (changes requested):
Hi,

Thanks for taking the time to put up a PR.

I am aware that Go has support for the loong64 arch, however, what systems exist that support loong64? Like aix, I am inclined to tell users to build the package themselves, rather than building for this arch on every single PR/nightly.

yzewei commented:
> Hi,
> 
> Thanks for taking the time to put up a PR.
> 
> I am aware that Go has support for the loong64 arch, however, what systems exist that support loong64? Like aix, I am inclined to tell users to build the package themselves, rather than building for this arch on every single PR/nightly.

thank you for your reply! Loong64 is about to get the upstream support of many distribution OS! There are already many available OS!
Please allow me to introduce below:
- Debian-sid has now provided Loongarch support [debian](https://wiki.debian.org/Ports/loong64)
- [Fedora](https://github.com/fedora-remix-loongarch/releases-info) and [Archlinux](https://mirrors.wsyu.edu.cn/loongarch/archlinux/iso/)'s Loongarch versions have already existed, as far as I know, I will apply to the upstream in the near future
- [Slackware](https://github.com/shipujin/slackware-loongarch64)
- [CLFS](https://github.com/sunhaiyong1978/CLFS-for-LoongArch)
- [Alt](https://www.altlinux.org/Ports/loongarch64)
- Alpine will add support after linux6.6
- There are also some server -side operating systems that also provide support to Loongarch, such as OpenUler, which belongs to Openatom; Openanolis, Kylin, UOS, etc.

yzewei commented:
@powersj What causes the CI failure?

powersj commented on Makefile:
```diff
@@ -428,6 +432,9 @@ mipsel.deb linux_mipsel.tar.gz: export GOARCH := mipsle
 riscv64.deb riscv64.rpm linux_riscv64.tar.gz: export GOOS := linux
 riscv64.deb riscv64.rpm linux_riscv64.tar.gz: export GOARCH := riscv64
 
+riscv64.deb riscv64.rpm linux_loongarch64.tar.gz: export GOOS := linux
+riscv64.deb riscv64.rpm linux_loongarch64.tar.gz: export GOARCH := loong64
```
I think you copy and pasted here, but did not update the riscv64 to loongarch64 in all places.

However, let's keep this to the tar.gz for now.

yzewei replied:
It may be difficult to verify in a physical machine, but there are several ways to simulate Loongarch on the X86_64 physical machine：
1. You can use it on x86 machine with docker run -it --rm merore/archlinux-loongarch64.
2. Use [qemu-user-static](https://hub.docker.com/r/multiarch/qemu-user-static)
The above two methods use the different forms of QEMU to simulate. The specific use method is on the official website. Of course, if you want to know how to use any of the above methods, I also have some shallow experience. We can communicate with each other.
About merore/archlinux-loongarch64:
- Boot an Loongarch64-based Arch Linux with qemu-system-loongarch64
- The default user/password is root/loongarch64
- All dependcies required by criu build and test are pre-installed
- You must configure your own dependence on the environment, Golang's installation package is located at http://ftp.loongnix.cn/toolchain/golang/go-1.20/abi2.0/go1.20.2.linux-loong64.tar.gz
About qemu-user-static:
- Loongarch has received the official support of QEMU, so you can download [qemu-loongarch-static](https://github.com/multiarch/qemu-user-static/releases/download/v7.2.0-1/qemu-loongarch64-static)
- If you want to use the Golang image of Loongarch, you can access lcr.looongnix.cn
-  If you want to build it yourself, lcr.loongnix.cn/library/debian:sid is the most suitable basic image

Finally, thank you for your patience !

yzewei replied:
Ok , thank you very much! 😊

powersj reviewed (changes requested):
@yzewei,

Talked to the team today about this further. While there is upstream software support, the hardware support is extremely limited. This is not something that I could go out and buy. That makes this much more difficult to support and justify.

I also need to be conscious of costs to us for the additional artifacts. There is no free lunch here.

As such I think we will tentatively accept a PR that adds builds to nightly and releases for now, but not to every single PR.

powersj commented:
@yzewei I would suggest updating your branch to the latest master to resolve some of the CI issues.

yzewei commented:
#https://app.circleci.com/pipelines/github/yzewei/telegraf/5/workflows/20be02c5-d5e7-4a79-b220-bca843fcf957/jobs/78?invite=true#step-109-20
Looks like a network problem

yzewei commented:
**test-go-mac**
Error Message:
 Resource class macos for macos.m1.medium.gen1, image xcode:14.2.0 is not available for your project, or is not a valid resource class. This message will often appear if the pricing plan for this project does not support macos use.

yzewei commented:
ptal

powersj commented:
I have updated your PR to remove building loong64 for every PR as I stated early. This should only build for new releases (tags) and for nightly. I've also updated the nightly download page to include the loong64 packages.

yzewei commented:
> I have updated your PR to remove building loong64 for every PR as I stated early. This should only build for new releases (tags) and for nightly. I've also updated the nightly download page to include the loong64 packages.
> 
> Again if loong64 ever holds up a release, we will drop support immediately.

Thank you for your help, and support loongarch64 to make many lovers using this architecture more convenient to use telegraf
